### PR TITLE
[FLINK-36318] Fix deserializing from 1.18 savepoint with RAW<MAP...> types

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

The commit fixes restoring from savepoints which were created for SQL queries which used functions such as LAG/LEAD/ARRAY_AGG which use a RAW type for accumulators.

## Verifying this change

Added tests in `org.apache.flink.table.planner.plan.nodes.exec.stream.OverWindowRestoreTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (**yes** / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 1]
```diff
@@ -146,144 +144,93 @@ protected Class<?> resolveProxyClass(String[] interfaces)
...
-     * are hardcoded and fixed.
-     *
-     * <p>IMPORTANT: mappings can be removed after 1 release as there will be a "migration path". As
-     * an example, a serializer is removed in 1.5-SNAPSHOT, then the mapping should be added for
-     * 1.5, and it can be removed in 1.6, as the path would be Flink-{< 1.5} -> Flink-1.5 ->
-     * Flink-{>= 1.6}.
+     * An {@link ObjectInputStream} that ignores certain serialVersionUID mismatches. This is a
+     * workaround for uncontrolled serialVersionUIDs changes.
      */
-    private enum MigrationUtil {
```
Comment (dawidwys) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java:250]: Those were unused. It was missed during previous clean ups.

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 2]
```diff
@@ -146,144 +144,93 @@ protected Class<?> resolveProxyClass(String[] interfaces)
...
-     */
-    private static final Set<String> scalaTypes = new HashSet<>();
+        void addVersionsMatch(
+                String className, long localVersionUID, List<Long> streamVersionUIDs) {
+            supportedSerialVersionUidsPerClass
+                    .computeIfAbsent(className, k -> new HashMap<>())
+                    .put(localVersionUID, streamVersionUIDs);
+        }
 
-    static {
```
Comment (dawidwys) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java:177]: Those were unused. It was missed during previous clean ups.

**Review by snuyanzin — APPROVED** [review]

**Comment by snuyanzin** [comment]
seems need to apply spotless

**Review by snuyanzin — APPROVED** [review]

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java [hunk 3]
```diff
@@ -144,7 +150,42 @@ private Stream<Arguments> createSpecs() {
...
+     * The method can be overridden in a subclass to test multiple savepoint files for a given
+     * program and a node in a particular version. This can be useful e.g. to test a node against
+     * savepoint generated in different Flink versions.
+     */
+    protected Stream<String> getSavepointPaths(
+            TableTestProgram program, ExecNodeMetadata metadata) {
+        return Stream.of(getSavepointPath(program, metadata, null).toString());
+    }
+
+    protected final Path getSavepointPath(
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java:177]: nit: return `String` to match with `getSavepointPaths`. also avoid the additional toString in the `OverWindowRestoreTest `

Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java:177]: the `Paths.get` can be done in `generateTestSetupFiles `

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/RowDataSerializer.java [hunk 4]
```diff
@@ -314,7 +314,8 @@ public void readSnapshot(int readVersion, DataInputView in, ClassLoader userCode
             types = new LogicalType[length];
             for (int i = 0; i < length; i++) {
                 try {
-                    types[i] = InstantiationUtil.deserializeObject(stream, userCodeClassLoader);
+                    types[i] =
+                            InstantiationUtil.deserializeObject(stream, userCodeClassLoader, true);
```
Comment (twalthr) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/RowDataSerializer.java:318]: Is RowDataSerializer enough? Wouldn't it be safer to do it for all constructed types? So MapDataSerializer and ArrayDataSerializer?

Comment (dawidwys) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/RowDataSerializer.java:318]: Very good point!

**Review by twalthr — APPROVED** [review]
Thanks for the update. LGTM

**Pull request closed without merging**